driver/resthandler.go

```diff
@@ -0,0 +1,364 @@
+		return
+	}
+
+	deviceResource, ok := handler.service.DeviceResource(deviceName, resourceName, "get")
+	if !ok {
+		handler.logger.Error(fmt.Sprintf("Incoming reading ignored. Resource '%s' not found", resourceName))
+		http.Error(writer, fmt.Sprintf("Resource '%s' not found", resourceName), http.StatusNotFound)
+		return
+	}
+
+	var reading interface{}
+	readingType := models.ParseValueType(deviceResource.Properties.Value.Type)
+
+	if readingType == models.Binary {
+		reading, err = handler.readBodyAsBinary(writer, request)
+	} else {
+		reading, err = handler.readBodyAsString(writer, request)
+	}
+
+	if err != nil {
+		handler.logger.Error(fmt.Sprintf("Incoming reading ignored. Unable to read request body: %s", err.Error()))
+		http.Error(writer, err.Error(), http.StatusBadRequest)
+		return
+	}
+
+	value, err := handler.newCommandValue(resourceName, reading, readingType)
+	if err != nil {
+		handler.logger.Error(fmt.Sprintf("Incoming reading ignored. Unable to create Command Value for Device=%s Command=%s: %s", deviceName, resourceName, err.Error()))
+		http.Error(writer, err.Error(), http.StatusBadRequest)
+		return
+	}
+
+	asyncValues := &models.AsyncValues{
+		DeviceName:    deviceName,
+		CommandValues: []*models.CommandValue{value},
+	}
+
+	handler.logger.Debug(fmt.Sprintf("Incoming reading received: Device=%s Resource=%s", deviceName, resourceName))
+
+	handler.asyncValues <- asyncValues
+}
+
+func (handler RestHandler) readBodyAsString(writer http.ResponseWriter, request *http.Request) (string, error) {
+	defer request.Body.Close()
+	body, err := ioutil.ReadAll(request.Body)
+	if err != nil {
+		return "", err
+	}
+
+	if len(body) == 0 {
+		return "", fmt.Errorf("no request body provided")
+	}
+
+	return string(body), nil
+}
+
+func (handler RestHandler) readBodyAsBinary(writer http.ResponseWriter, request *http.Request) ([]byte, error) {
+	defer request.Body.Close()
+	body, err := ioutil.ReadAll(request.Body)
+	if err != nil {
+		return nil, err
+	}
+
+	if len(body) == 0 {
+		return nil, fmt.Errorf("no request body provided")
+	}
+
+	return body, nil
+}
+
+func deviceHandler(writer http.ResponseWriter, request *http.Request) {
+	handler, ok := request.Context().Value(handlerContextKey).(RestHandler)
+	if !ok {
+		writer.WriteHeader(http.StatusBadRequest)
+		writer.Write([]byte("Bad context pass to handler"))
+		return
+	}
+
+	handler.processAsyncRequest(writer, request)
+}
+
+func (handler RestHandler) newCommandValue(resourceName string, reading interface{}, valueType models.ValueType) (*models.CommandValue, error) {
+	var result = &models.CommandValue{}
+	var err error
+	var timestamp = time.Now().UnixNano()
+	castError := "fail to parse %v reading, %v"
+
+	if !checkValueInRange(valueType, reading) {
+		err = fmt.Errorf("parse reading fail. Reading %v is out of the value type(%v)'s range", reading, valueType)
+		handler.logger.Error(err.Error())
+		return result, err
+	}
+
+	switch valueType {
```

Comment:
Are the uses of NewCommandValue here correct in the light of comments in [https://github.com/edgexfoundry/device-sdk-go/issues/345](url)?